nanoPU_sim.py

```diff
@@ -519,49 +497,37 @@ def start(self):
                 self.buffers[tx_msg_id].append(msg[(num_pkts-1)*Simulator.max_pkt_len:])
                 # initialize other state
                 self.delivered[tx_msg_id] = 0
-                self.credit[tx_msg_id] = Simulator.rtt_pkts - 1 # index of 10th pkt is 9
+                self.credit[tx_msg_id] = Simulator.rtt_pkts
                 self.toBtx[tx_msg_id] = (1<<num_pkts)-1 # every pkt must be transmitted
                 self.max_tx_pkt_offset[tx_msg_id] = 0
                 self.timeout_count[tx_msg_id] = 0
                 # schedule a timer for this msg
                 self.scheduleTimerEvent(tx_msg_id, 0)
-                # make this message active
-                self.enq_active_messages_fifo(tx_msg_id)
+                # schedule the first pkts of the msg for transmission
+                tx_pkts = self.toBtx[tx_msg_id] & (1<<self.credit[tx_msg_id])-1
+                self.scheduled_pkts_fifo.put((tx_msg_id, tx_pkts))
             else:
                 self.log('ERROR: dropping message due to lack of an available tx_msg_id')
 
-    def enq_active_messages_fifo(self, tx_msg_id):
-        # make sure msg is not already active
-        if self.active_messages_bitmap & (1<<tx_msg_id) == 0:
-            self.log('Enqueueing msg {} into active_messages_fifo'.format(tx_msg_id))
-            self.active_messages_fifo.put(tx_msg_id)
-            # mark as active (flip bit)
-            self.active_messages_bitmap = self.active_messages_bitmap ^ (1<<tx_msg_id)
-        else:
-            self.log('Msg {} is already active')
-
     def dequeue(self):
         """Send a data pkt to the arbiter. The arbiter invokes this method when it decides to
            schedule a data pkt.
         """
-        wait_active_msg_event = self.active_messages_fifo.get()
-        try:
-            # wait for a msg to become active
-            tx_msg_id = yield wait_active_msg_event
-        except simpy.Interrupt as i:
-            wait_active_msg_event.cancel()
-            self.env.exit(None)
-
-        # mark msg as inactive (flip bit)
-        self.active_messages_bitmap = self.active_messages_bitmap ^ (1<<tx_msg_id)
-        # lookup which pkts of this msg need to be transmitted
-        toBtx_bitmap = self.toBtx[tx_msg_id]
-        credit = self.credit[tx_msg_id]
-        credit_bitmap = (1<<credit)-1
-        # the only pkts that are eligible are the ones at an index < credit
-        eligible_pkts_bitmap = toBtx_bitmap & credit_bitmap
+        # check if we've finished sending all the currently active pkts
+        if self.active_tx_pkts == 0:
```

Comment:
We are transferring all the permitted packets of a message before handling another message. This is definitely a legitimate option. However this means that two messages become ready to send 10 packets at the same time, one of them send its 10 packets in a single burst and the other waits for 10 packet serialization times before being able to send any packet. I think this increases burstiness and promotes more congestion in the network.
Especially for nanoPU world where most of the messages are short so that they would be sent as bursts (creating incast) the natural pacing (msg handling in round robin fashion) could be somewhat helpful. What do you think?

Reply:
In what situations would something like this happen? i.e. a PULL pkt arrives and makes a whole bunch of pkts available for tx? Maybe if the PULLs are significantly reordered or some are dropped in the network? In the usual case each PULL should only trigger generation of 1 pkt, right?

Reply:
The examples you give are possible. 
Yes ideally a PULL should generate only a single packet but this is not guaranteed and not rare enough because of reordering.
In addition, imagine CPU writes 2 messages to the packetization buffer back to back. Let's say by the time the CPU completes writing the second message, the first messages has not been dequed to the egress pipeline yet because of already existing messages in the tx_fifo. Then the second message will not be able to send any packets at all until the first message sends out 10 packets into the network.

Reply:
I agree PULL pkts may generate some bursts of multiple pkts from the same msg. However, the way that this behavioral model implements the scheduling of tx pkts for msg writes from the CPU does not accurately represent the HW implementation. After the CPU finishes writing each pkt, the msg processing thread will insert an entry into the scheduled_pkts_fifo. The creditToBtxEvent thread may insert entries between these writes, so there won't necessarily be a burst of pkts from the same msg.

I don't think it's worth updating this behavioral model to be exactly like the HW implementation, there are diminishing returns.

Another edge case that comes up in the HW implementation but not here in the behavioral model is: what happens when a PULL pkt tries to pull a pkt that has not been written by the CPU yet? I think we may need another state variable (e.g. `available`) so that the creditToBtxEvent doesn't try to send an unavailable pkt...

Reply:
Alright. Makes sense.